### PR TITLE
fix(compiler): onLift not called, or called too many times

### LINKED ISSUE #5561: compile error when lifting the same resource multiple times
### I tried this:

this only happens when compiling to `tf-aws`:

```wing
bring cloud;

let bucket = new cloud.Bucket();

pub class MyBucket {
  bucket: cloud.Bucket;

  new(bucket: cloud.Bucket) {
    this.bucket = bucket;
  }

  pub inflight list() {
    this.bucket.list();
  }
}

let b1 = new MyBucket(bucket) as "b1";
let b2 = new MyBucket(bucket) as "b2";

let api = new cloud.Api();
api.get("/", inflight () => {
  b1.list();
  b2.list();
  return {
    status: 200
  };
});
```

### This happened:

```
eladcohen@Elads-w-MBP a2 % wing compile -t tf-aws main.w                    
runtime error: Environment variable "BUCKET_NAME_d755b447" already set with a different value.
  --> main.w:5:1
  | 
  | let bucket = new cloud.Bucket();
  | 
5 | pub class MyBucket {
  | ^
```

### I expected this:

_No response_

### Is there a workaround?

_No response_

### Anything else?

_No response_

### Wing Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #5573: onlift not beeing executed
### I tried this:

Use the websockets winglib with the provided example [here](https://github.com/winglang/winglibs/tree/main/websockets#usage).

### This happened:

In wing versions `0.54.38` and later:

```
{
    "errorType": "AccessDeniedException",
    "errorMessage": "User: arn:aws:sts::502852838290:assumed-role/terraform-20240129202613877600000001/on-message-c81978b1 is not authorized to perform: execute-api:ManageConnections on resource: arn:aws:execute-api:eu-west-3:********8290:vhllpy8kkb/prod/POST/@connections/{connectionId}",
    "name": "AccessDeniedException",
    "$fault": "client",
    "$metadata": {
        "httpStatusCode": 403,
        "requestId": "397ca1b4-919a-4d51-b831-7da4bd5b3508",
        "attempts": 1,
        "totalRetryDelay": 0
    },
    "message": "User: arn:aws:sts::502852838290:assumed-role/terraform-20240129202613877600000001/on-message-c81978b1 is not authorized to perform: execute-api:ManageConnections on resource: arn:aws:execute-api:eu-west-3:********8290:vhllpy8kkb/prod/POST/@connections/{connectionId}",
    "stack": [
        "AccessDeniedException: User: arn:aws:sts::502852838290:assumed-role/terraform-20240129202613877600000001/on-message-c81978b1 is not authorized to perform: execute-api:ManageConnections on resource: arn:aws:execute-api:eu-west-3:********8290:vhllpy8kkb/prod/POST/@connections/{connectionId}",
        "    at throwDefaultError (/node_modules/@smithy/smithy-client/dist-cjs/index.js:838:20)",
        "    at <anonymous> (/node_modules/@smithy/smithy-client/dist-cjs/index.js:847:5)",
        "    at de_PostToConnectionCommandError (/node_modules/@aws-sdk/client-apigatewaymanagementapi/dist-cjs/index.js:361:14)",
        "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
        "    at <anonymous> (/node_modules/@smithy/middleware-serde/dist-cjs/index.js:35:20)",
        "    at <anonymous> (/node_modules/@aws-sdk/middleware-signing/dist-cjs/index.js:184:18)",
        "    at <anonymous> (/node_modules/@smithy/middleware-retry/dist-cjs/index.js:320:38)",
        "    at <anonymous> (/node_modules/@aws-sdk/middleware-logger/dist-cjs/index.js:33:22)",
        "    at Object._postToConnection (/node_modules/@winglibs/websockets/inflight/websocket.aws.js:16:3)",
        "    at WebSocket_tfaws.sendMessage (/node_modules/@winglibs/websockets/platform/tf-aws.w:154:5)"
    ]
}
```

### I expected this:

_No response_

### Is there a workaround?

Use a version of wing prior to `0.54.38`.

### Anything else?

I think the onLift is not working properly after [this PR](https://github.com/winglang/wing/pull/5435).

websockets onlift code [here](https://github.com/winglang/winglibs/blob/main/websockets/platform/tf-aws.w#L46-L55)

### Wing Version

0.56.4

### Node.js Version

_No response_

### Platform(s)

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by Chriscbr**
This PR refactors code defining the interface between the compiler and core TypeScript library, the pieces of which are responsible for "lifting" objects and granting capabilities (like environment variables and IAM permissions) through the `onLift` and `onLiftType` methods. Through this refactoring we're able to fix a couple regressions that came out of #5435.

Today, the contract we’re trying to provide with the public `onLift()` method which can be added to any preflight class, is as follows. `onLift()` will be called on an object at most per inflight `host` argument.

As a concrete example, suppose you write some code like this in your class:

```js
// ... 
class A {
  onLift(host: std.IInflightHost, ops: Array<str>) {
    log(“I was lifted by “ + host.node.path);
  }
}
// …
```

Then you should never see the same line printed two or more times:
```
$ wing compile main.w
I was lifted by root/cloud.Function
I was lifted by root/cloud.Function
```

The motivation is this that it’s a better DX for library authors. The main utility of `onLift` is to add mutations or additions to the `host`. The idempotency requirement ensures you don’t have to defensively code as if the method will be called multiple times with the same host.

The implementation of this has been tricky though for various reasons (like lazy execution, and the fact that we have to deal with class inheritance). In part, #5435 introduce a regression where this contract was violated.

At a high level, we've fixed this by replacing an internal utility method, `onLiftMatrix()`, (which partially de-duplicated `onLift` calls) with a separate algorithm that traverses an object dependency graph based the information we're generating with the compiler. The core logic is implemented in lifting.ts (https://github.com/winglang/wing/blob/43d7864d3509dd08085da2835dfacf36783318d2/libs/wingsdk/src/core/lifting.ts).
A collection of objects and each of the operations requested on them is generated. Once we have this collection, only then do we call `onLift` and `onLiftType` methods.

Several tests have been added to validate edge cases where de-duplication issues arise.

Fixes #5561
Fixes #5573

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingsdk/src/cloud/function.ts [hunk 1]
```diff
@@ -100,14 +100,14 @@ export class Function extends Resource implements IInflightHost {
...
-    // `require` resolves naturally.
+    // write the entrypoint next to the partial inflight code emitted by the compiler,
+    // so that `require` resolves naturally.
     const lines = this._getCodeLines(this.handler);
     writeFileSync(this.entrypoint, lines.join("\n"));
 
     // indicates that we are calling the inflight constructor and the
     // inflight "handle" method on the handler resource.
-    this.handler.onLift(this, ["handle", "$inflight_init"]);
+    Lifting.lift(this.handler, this, ["handle"]);
```
Comment (Chriscbr) [inline libs/wingsdk/src/cloud/function.ts:110]: (half baked thoughts) Eventually I imagine this lifting thing will need to be a public API. But right now it's called in this `_preSynthesize` method, and based on past issues dealing with `preSynthesize` bugs in projen, I'd like to avoid exposing the capability for users to execute arbitrary code in this lifecycle hook.

Some possible solutions (not planning to implement in this PR) could be:
1. Change the API so that `Lifting.lift` can be called inside your class constructor, but it only does its actual work during `preSynthesize` under the hood
2. Add a special base class for compute resources with some built-in methods you can use, e.g.

```js
bring cloud;

class Function extends std.InflightHost {
  new() {
    // ...
    super.lift(this.handler, ["handle"]);
    // ...
  }
}
```

Anyway I think we have a lot of options for how to handle this

File: libs/wingsdk/src/std/resource.ts [hunk 2]
```diff
@@ -80,6 +80,14 @@ export interface ILiftable {
  * to access the lifted object inflight.
  */
 export interface IHostedLiftable extends ILiftable {
+  /**
+   * Compiler-generated data that describes the dependencies of this object on other
+   * objects. This is used to determine which permissions need to be granted to the
+   * inflight host.
+   * @internal
+   */
+  _onLiftDeps?: LiftDepsMatrixRaw;
```
Comment (Chriscbr) [inline libs/wingsdk/src/std/resource.ts:89]: Wasn't sure what to name this internal property. I guess since it's internal it doesn't matter too much, but suggestions are still welcome

File: examples/tests/valid/lift_weird_order.test.w
Comment (Chriscbr) [inline examples/tests/valid/lift_weird_order.test.w:1]: I can move this to be a unit test in the SDK

File: libs/@wingcloud/framework/src/inflight.ts [hunk 3]
```diff
@@ -161,6 +161,18 @@ class Lifter<
       ...args: Parameters<TFunction>
     ) => ReturnType<TFunction>
   ): Inflight<TFunction> {
+    const _onLiftDeps: LiftDepsMatrixRaw = { handle: [], $inflight_init: [] };
+    for (const [key, obj] of Object.entries(this.lifts)) {
```
Comment (eladb) [inline libs/@wingcloud/framework/src/inflight.ts:165]: Maybe add a wee comment explaining what's going on here

File: libs/wingc/src/jsify.rs [hunk 4]
```diff
@@ -1795,34 +1777,61 @@ impl<'a> JSifier<'a> {
...
-					.kind;
-				let is_static = matches!(var_kind, VariableKind::StaticMember);
-				(*m == CLASS_INFLIGHT_INIT_NAME || !is_static) ^ (matches!(bind_method_kind, BindMethod::Type))
-			})
-			.collect_vec();
-
-		// Skip jsifying this method if there are no lifts (in this case we'll use super's register bind method)
+		let mut lift_qualifications: Vec<(&String, &BTreeMap<String, LiftQualification>)> = vec![];
+		let empty_map = BTreeMap::new();
+
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1782]: Add some comments here as well

File: libs/wingc/src/jsify/snapshots/base_class_captures_preflight.snap [hunk 5]
```diff
@@ -118,8 +121,11 @@ class $Root extends $stdlib.std.Resource {
           })())
         `;
       }
-      _supportedOps() {
```
Comment (eladb) [inline libs/wingc/src/jsify/snapshots/base_class_captures_preflight.snap:121]: Just to make sure I understand, so now onLiftDeps is used to determine supportedOps ?

File: libs/wingc/src/dtsify/snapshots/declarations.snap [hunk 6]
```diff
@@ -172,8 +172,21 @@ class ParentClass extends $stdlib.std.Resource {
       })())
     `;
   }
-  _supportedOps() {
-    return [...super._supportedOps(), "static_method", "foo", "bar", "$inflight_init"];
+  get _onLiftDeps() {
```
Comment (eladb) [inline libs/wingc/src/dtsify/snapshots/declarations.snap:175]: The `on` prefix is confusing.. since this is a property maybe just call this `_liftDeps` or `_lifts` or `_liftMap` or something along those lines...

**Review by MarkMcCulloh — APPROVED** [review]

File: libs/wingc/src/jsify/snapshots/base_class_captures_preflight.snap (hunk 5 above)
Comment (Chriscbr) [reply libs/wingc/src/jsify/snapshots/base_class_captures_preflight.snap:121]: Yep. The idea I had was to replace `supportedOps` with `_liftDeps` since they're provide the same information

File: libs/@wingcloud/framework/src/inflight.ts (hunk 3 above)
Comment (Chriscbr) [reply libs/@wingcloud/framework/src/inflight.ts:170]: Added

File: libs/wingc/src/dtsify/snapshots/declarations.snap (hunk 6 above)
Comment (Chriscbr) [reply libs/wingc/src/dtsify/snapshots/declarations.snap]: Renamed to `_liftMap`

File: libs/wingc/src/jsify.rs [hunk 7]
```diff
@@ -1795,34 +1777,61 @@ impl<'a> JSifier<'a> {
...
+		for m in class.all_methods(true) {
+			let name = m.name.as_ref().unwrap();
+			let method = class_type.as_class().unwrap().get_method(&name);
+			let var_info = method.expect(&format!("method \"{name}\" doesn't exist in {class_name}"));
+			let var_kind = &var_info.kind;
+
+			let is_static = matches!(var_kind, VariableKind::StaticMember);
```
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingc/src/jsify.rs:1789]: Nice!

File: libs/wingsdk/src/core/lifting.ts [hunk 8]
```diff
@@ -90,153 +97,262 @@ export function liftObject(obj: any): string {
...
+    // Add the new ops to the explored set.
+    for (const op of ops) {
+      existingOps.add(op);
+    }
+
+    // Inspect the object to see if there are any transitive dependency information.
+    // Currently there are a few ways to do this:
+    // - The compiler may generate a _onLiftDeps property on the object
+    // - The compiler may generate a static _onLiftTypeDeps method on a class
+    // - The SDK may have a _supportedOps method on a class (TODO: remove this?)
```
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingsdk/src/core/lifting.ts:247]: It's not that big but it requires updating a lot of SDK code so I'd rather handle it in a separate PR

File: libs/wingsdk/test/target-sim/__snapshots__/file-counter.test.ts.snap [hunk 9]
```diff
@@ -56,21 +56,11 @@ bucket: (function() {
         "source": "root/HelloWorld/Queue/SetConsumer0",
         "target": "root/HelloWorld/Counter",
       },
-      {
-        "name": "$inflight_init()",
-        "source": "root/HelloWorld/Queue/SetConsumer0",
-        "target": "root/HelloWorld/Counter",
-      },
```
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingsdk/test/target-sim/__snapshots__/file-counter.test.ts.snap:63]: Yeah - I think these are OK to remove for now since the console ignores these connections
